### PR TITLE
"Shrouded" systems that can't be seen unless in visible range

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature:** This PR is a reopening of #5979.

## Feature Details
This PR adds a new system tag:
* `shrouded`: This system can't be seen unless it is within view range or linked to a system that can be viewed. Being visited is not a condition for a shrouded system to be seen, meaning the system is forgotten after you leave it. The only time you can actually view the contents of a shrouded system is when you are actually in it. Note though that the system will still count as visited for the purposes of maps and the visited condition.

As a reminder, hidden's definition is the following:
* `hidden`: This system can't be seen unless it is visited or linked to a visited system.

Combining hidden and shrouded then results in the following behavior:
* This system can't be seen unless it is linked to an unshrouded visited system.

## Usage Examples + Screenshots
I changed Sol, Alpha Centauri, and a number of other systems to be shrouded.
```
system Sol
	shrouded
	...
```
<details><summary>The following are a collection of screenshots of me flying through these shrouded systems. Observe how those systems that are shrouded and within range are seen, but the system name and their links to other shrouded system are forgotten after leaving.</summary>

image
image
image
image
</details>

<details><summary>I then added hidden to all these systems so that they are both shrouded and hidden. The behavior is mostly the same, except now I can't see those systems which are within range but not linked to any systems I can see.</summary>

image
image
</details>

## Testing Done
Nothing on this version. I'll test it later. It sure did compile, though. (Will undraft it when that's done.)

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

File: source/MapDetailPanel.cpp [hunk 1]
```diff
@@ -723,7 +723,7 @@ void MapDetailPanel::DrawInfo()
 	font.Draw({systemName, alignLeft}, uiPoint + Point(0., -7.), medium);
 
 	governmentY = uiPoint.Y() + textMargin;
-	string gov = player.HasVisited(*selectedSystem) ?
+	string gov = player.CanView(*selectedSystem) ?
```
Comment (RisingLeaf) [inline source/MapDetailPanel.cpp:726]: why not use the variable from earlier?

File: source/MapDetailPanel.cpp [hunk 2]
```diff
@@ -752,8 +752,8 @@ void MapDetailPanel::DrawInfo()
 
 		string price;
 
-		bool hasVisited = player.HasVisited(*selectedSystem);
-		if(hasVisited && selectedSystem->IsInhabited(player.Flagship()))
+		bool canView = player.CanView(*selectedSystem);
```
Comment (Amazinite) [inline source/MapDetailPanel.cpp:755]: This one is redundant too.
```suggestion
```

**Comment by Amazinite** [comment]
This is working as intended now. The remaining bug we found on stream turned out to be a vanilla issue. I've opened #9624 to fix it.

File: source/DistanceMap.cpp [hunk 3]
```diff
@@ -249,15 +249,19 @@ void DistanceMap::Init(const Ship *ship)
 
 					// In order to plan travel through a wormhole, it must be
 					// "accessible" to your flagship, and you must have visited
-					// the wormhole and both endpoint systems. (If this is a
-					// multi-stop wormhole, you may know about some paths that
-					// it takes but not others.)
+					// the wormhole and both endpoint systems must be viewable.
+					// (If this is a multi-stop wormhole, you may know about
+					// some paths that it takes but not others.)
+					if(ship && !object.GetPlanet()->IsAccessible(ship))
```
Comment (Amazinite) [inline source/DistanceMap.cpp:255]:
```suggestion
```

File: source/Engine.cpp [hunk 4]
```diff
@@ -642,7 +642,6 @@ void Engine::Step(bool isActive)
 	if(flagship)
 		ammoDisplay.Update(*flagship);
```
Comment (Amazinite) [inline source/Engine.cpp:644]:
```suggestion


```

File: source/PlayerInfo.cpp [hunk 5]
```diff
@@ -2393,18 +2398,41 @@ bool PlayerInfo::HasSeen(const System &system) const
...
+		return m.Destination()->IsInSystem(&system);
 	};
 	if(any_of(availableJobs.begin(), availableJobs.end(), usesSystem))
 		return true;
 	if(any_of(missions.begin(), missions.end(), usesSystem))
 		return true;
 
+	if(shrouded)
+	{
+		// All systems linked to a visible system are themselves visible.
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:2410]:
```suggestion
		// All systems linked to a system the player can view are visible.
```

File: source/PlayerInfo.cpp [hunk 6]
```diff
@@ -2393,18 +2398,41 @@ bool PlayerInfo::HasSeen(const System &system) const
...
+		if(any_of(system.Links().begin(), system.Links().end(),
+				[&](const System *s) noexcept -> bool { return CanView(*s); }))
+			return true;
+		// A shrouded system not linked to a visible system must be within view range to be seen.
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:2414]:
```suggestion
		// A shrouded system not linked to a viewable system must be visible from the current system.
```

**Review by RisingLeaf — COMMENTED** [review]
Looks good to me, only thing, shouldn't you also replace hasVisited with can see in distancemap 339?

**Comment by Amazinite** [comment]
Good catch.

**Review by RisingLeaf — APPROVED** [review]

**Pull request merged**